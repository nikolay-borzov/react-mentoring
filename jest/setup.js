import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
