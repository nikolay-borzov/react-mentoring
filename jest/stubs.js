export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'http://via.placeholder.com/500x750',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre']
}

export const films = [film]
