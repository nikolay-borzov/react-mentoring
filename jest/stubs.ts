import { Film } from '../src/entities/film'

export const film: Film = {
  id: 1,
  title: 'Film title',
  tagline: 'Put that cookie down',
  vote_average: 7.5,
  vote_count: 667,
  release_date: '2018-05-04',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Once upon a time...',
  budget: 15000000,
  revenue: 59735548,
  genres: ['some genre', 'another genre']
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id: number) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
