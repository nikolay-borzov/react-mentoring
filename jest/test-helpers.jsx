import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
