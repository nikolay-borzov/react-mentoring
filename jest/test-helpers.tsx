import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
