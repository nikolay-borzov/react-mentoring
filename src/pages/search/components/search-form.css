.search-form-container {
  margin-top: 1rem;
}