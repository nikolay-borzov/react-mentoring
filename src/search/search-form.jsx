import React from 'react'

export class SeachForm extends React.PureComponent {
  render() {
    return (
      <div className="search-form-container alt-background">
        <form name="search-form">
          <div className="form-row">
            <label className="form-label">Find your movie</label>
          </div>
          <div className="form-row">
            <input type="text" className="text-input" />
          </div>
          <div className="form-row">
            <label className="form-label">Search by</label>
            <div className="radio-input">
              <input
                type="radio"
                id="searchBy_title"
                name="searchBy"
                value="title"
              />
              <label
                htmlFor="searchBy_title"
                className="radio-input__label button button--small">
                Title
              </label>
            </div>
            <div className="radio-input">
              <input
                type="radio"
                id="searchBy_director"
                name="searchBy"
                value="director"
              />
              <label
                htmlFor="searchBy_director"
                className="radio-input__label button button--small">
                Director
              </label>
            </div>
            <div className="flex-grow align-right">
              <input
                type="submit"
                className="button button--primary"
                value="Search"
              />
            </div>
          </div>
        </form>
      </div>
    )
  }
}
