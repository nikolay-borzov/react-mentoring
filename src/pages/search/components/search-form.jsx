// @flow

import * as React from 'react'
import styled from 'styled-components'

import { searchBy } from '../../../enums'

import { Radio } from '../../../components'
import { Button, FormRow, FormLabel } from '../../../styles'

type SearchFormProps = {
  search: string,
  searchBy: string,
  onSearchChange: ({ search: string, searchBy: string }) => void
}

const SearchFormContainer = styled.div`
  margin-top: 1rem;
`

const SearchInput = styled.input`
  width: 100%;
  font-family: 'Regular';
  font-size: 1.3rem;
  background-color: var(--color-bc-alt);
  border: none;
  border-bottom: 2px solid var(--color-primary);
  padding: var(--padding-input);
  color: var(--color-text-light);
`

export class SearchForm extends React.PureComponent<SearchFormProps> {
  searchByOptions = [
    {
      name: 'Title',
      value: searchBy.title
    },
    {
      name: 'Genre',
      value: searchBy.genres
    }
  ]

  searchInput: ?HTMLInputElement
  searchByInput: ?Radio

  constructor(props) {
    super(props)

    this.searchInput = React.createRef()
    this.searchByInput = React.createRef()
  }

  onSubmit = (event: SyntheticEvent<HTMLFormElement>) => {
    event.preventDefault()

    const searchInput = this.searchInput.current
    const searchByInput = this.searchByInput.current

    if (searchInput && searchByInput) {
      this.props.onSearchChange({
        search: searchInput.value,
        searchBy: searchByInput.value
      })
    }
  }

  render() {
    const { search, searchBy } = this.props

    return (
      <SearchFormContainer>
        <form name="search-form" onSubmit={this.onSubmit}>
          <FormRow>
            <FormLabel className="uppercase">Find your movie</FormLabel>
          </FormRow>

          <FormRow>
            <SearchInput
              name="search"
              innerRef={this.searchInput}
              type="text"
              defaultValue={search}
            />
          </FormRow>

          <FormRow className="uppercase">
            <Radio
              ref={this.searchByInput}
              name="searchBy"
              label="Search by"
              defaultValue={searchBy}
              options={this.searchByOptions}
              style="button"
            />

            <div className="flex-grow align-right">
              <Button type="submit" primary value="Search" />
            </div>
          </FormRow>
        </form>
      </SearchFormContainer>
    )
  }
}
