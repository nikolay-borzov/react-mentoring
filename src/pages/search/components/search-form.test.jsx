import React from 'react'

import { SearchForm } from './search-form'

import { itRendersCorrectlyShallow } from '../../../../jest/test-helpers'

describe('SearchForm component', () => {
  let props = {
    onSearchChange: jest.fn()
  }

  beforeEach(() => {
    props.search = ''
    props.searchBy = 'title'

    props.onSearchChange.mockClear()
  })

  itRendersCorrectlyShallow(() => <SearchForm {...props} />)

  it(`calls 'onSearchChange' on form submit`, () => {
    props.search = 'A movie'
    const wrapper = mount(<SearchForm {...props} />)
    const eventMock = {
      preventDefault: jest.fn()
    }

    wrapper.find('form').simulate('submit', eventMock)

    expect(eventMock.preventDefault).toHaveBeenCalled()
    expect(props.onSearchChange).toHaveBeenCalledWith({
      search: props.search,
      searchBy: props.searchBy
    })
  })
})
