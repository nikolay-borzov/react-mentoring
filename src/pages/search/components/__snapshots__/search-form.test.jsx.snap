// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm component renders correctly 1`] = `
<div
  className="search-form-container"
>
  <form
    name="search-form"
    onSubmit={[Function]}
  >
    <div
      className="form-row"
    >
      <label
        className="form-label uppercase"
      >
        Find your movie
      </label>
    </div>
    <div
      className="form-row"
    >
      <input
        className="text-input"
        defaultValue=""
        name="search"
        type="text"
      />
    </div>
    <div
      className="form-row uppercase"
    >
      <Radio
        defaultValue="title"
        label="Search by"
        name="searchBy"
        options={
          Array [
            Object {
              "name": "Title",
              "value": "title",
            },
            Object {
              "name": "Genre",
              "value": "genres",
            },
          ]
        }
        style="button"
      />
      <div
        className="flex-grow align-right"
      >
        <input
          className="button button--primary"
          type="submit"
          value="Search"
        />
      </div>
    </div>
  </form>
</div>
`;
